test/fixtures/azure-fixture/src/main/java/fixture/azure/MockAzureBlobStore.java

```diff
@@ -0,0 +1,483 @@
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.rest.RestStatus;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public class MockAzureBlobStore {
+
+    private static final Logger logger = LogManager.getLogger(MockAzureBlobStore.class);
+    private static final BytesReference EMPTY = new BytesArray(new byte[] {});
+
+    public enum BlobType {
+        BLOCK("BlockBlob") {
+            @Override
+            AzureBlob create() {
+                return new MockAzureBlockBlob();
+            }
+        },
+        PAGE("PageBlob") {
+            @Override
+            AzureBlob create() {
+                throw new UnsupportedOperationException("PageBlob is not supported");
+            }
+        },
+        APPEND("AppendBlob") {
+            @Override
+            AzureBlob create() {
+                throw new UnsupportedOperationException("AppendBlob is not supported");
+            }
+        };
+
+        final String xMsBlobType;
+
+        BlobType(String xMsBlobType) {
+            this.xMsBlobType = xMsBlobType;
+        }
+
+        abstract AzureBlob create();
+
+        public String getXMsBlobType() {
+            return xMsBlobType;
+        }
+
+        public static BlobType fromXMSBlobType(String blobTypeString) {
+            for (BlobType blobType : BlobType.values()) {
+                if (blobType.xMsBlobType.equals(blobTypeString)) {
+                    return blobType;
+                }
+            }
+            return null;
+        }
+    }
+
+    private final Map<String, AzureBlob> blobs;
+
+    public MockAzureBlobStore() {
+        this.blobs = new ConcurrentHashMap<>();
+    }
+
+    public void putBlock(String path, String blockId, BytesReference content, @Nullable String leaseId) {
+        blobs.compute(path, (p, existing) -> {
+            if (existing != null) {
+                if (existing instanceof MockAzureBlockBlob mabb) {
+                    mabb.putBlock(blockId, content, leaseId);
+                    return existing;
+                } else {
+                    throw new ConflictException("InvalidBlobType", "The blob type is invalid for this operation.");
+                }
+            } else {
+                final MockAzureBlockBlob mockAzureBlockBlob = new MockAzureBlockBlob();
+                mockAzureBlockBlob.putBlock(blockId, content, leaseId);
+                return mockAzureBlockBlob;
+            }
+        });
+    }
+
+    public void putBlockList(String path, List<String> blockIds, @Nullable String leaseId) {
+        final AzureBlob azureBlob = getExistingBlob(path);
+        if (azureBlob instanceof MockAzureBlockBlob mabb) {
+            mabb.putBlockList(blockIds, leaseId);
+        } else {
+            throw new ConflictException("InvalidBlobType", "The blob type is invalid for this operation.");
```

Comment:
On paths we expect never to follow, could we fail the test rather than just returning a valid error response to ES? That way it's clearer that we aren't exercising these paths and relying on a particular error response. Probably needs `org.elasticsearch.ExceptionsHelper#maybeDieOnAnotherThread` since there's a `catch (Throwable t)` somewhere in the HTTP server stack that will turn even an `AssertionError` into a 500 unless we throw it on a different thread.

Reply:
I have applied that logic to all the paths that aren't implemented (i.e. attempting `renew`/`break`/`change` a lease, or attempting to create a `PageBlob` or `AppendBlob`). But this particular case I wouldn't mind keeping as-is. We don't expect to take the branch, and we won't unless someone implements another blob type, at which point the behaviour here will be correct?

Reply:
> We don't expect to take the branch, and we won't unless someone implements another blob type, at which point the behaviour here will be correct?

This is kind of my point: we don't really know how it should behave in some future situation that we haven't encountered yet. I mean it sounds to me like a bug to try and `putBlockList` on a non-block blob, maybe we'd want it to fail the test in that case too? Let's not do anything other than fail the test on branches we believe to be dead code.